post-processor/vsphere-template/step_mark_as_template.go

```diff
@@ -23,43 +21,31 @@ func (s *stepMarkAsTemplate) Run(state multistep.StateBag) multistep.StepAction
 	folder := state.Get("folder").(*object.Folder)
 	dcPath := state.Get("dcPath").(string)
 
-	ui.Message("Marking as a template...")
-
-	vm, err := findRuntimeVM(cli, dcPath, s.VMName)
-	if err != nil {
+	ui.Message(fmt.Sprintf("Unregistering template %s/%s", folder.InventoryPath, s.VMName))
+	if err := unregisterPreviousVM(cli, folder, s.VMName); err != nil {
 		state.Put("error", err)
 		ui.Error(err.Error())
 		return multistep.ActionHalt
 	}
 
-	host, err := vm.HostSystem(context.Background())
+	ui.Message(fmt.Sprintf("Finding vm %s", s.VMName))
+	vm, err := findRuntimeVM(cli, dcPath, s.VMName)
 	if err != nil {
 		state.Put("error", err)
 		ui.Error(err.Error())
 		return multistep.ActionHalt
 	}
 
-	if err := vm.Unregister(context.Background()); err != nil {
-		state.Put("error", err)
-		ui.Error(err.Error())
-		return multistep.ActionHalt
-	}
-
-	source := strings.Split(s.Source, "/vmfs/volumes/")[1]
-	i := strings.Index(source, "/")
-
-	path := (&object.DatastorePath{
-		Datastore: source[:i],
-		Path:      source[i:],
-	}).String()
-
-	if err := unregisterPreviousVM(cli, folder, s.VMName); err != nil {
+	ui.Message(fmt.Sprintf("Marking as template %s", s.VMName))
+	err = vm.MarkAsTemplate(context.Background())
```

Comment:
Sadly you only can use `vm.MarkAsTemplate(context.Background())` when the status VM is registered, the only way to maintain this status is setting the attribute `keep_registered` as true in the packer template. This is why the original implementation uses `RegisterVM` instead.

As `keep_registered` is false by [default](https://www.packer.io/docs/builders/vmware-iso.html#keep_registered), this fix will bring us more issues.

Reply:
Totally agree, I've completely missed that point. Probably, I can tune it to check the builder properties and detect if keep_registered is true. Otherwise, I think we should move forward with the solution proposed under PR #5380, but IMHO, having the vm already in remote vpshere takes less time to simple mark it as template, but yes, it must be generated in an ESXi, and probably users has build the image using another support like vmplayer. @bugbuilder thanks for checking this!

Reply:
Thanks for your feedback and your work!